### PR TITLE
THREESCALE-5661 upgrade formtastic

### PR DESCRIPTION
**Pull request by nidhi-soni1104**
**What this PR does / why we need it**:
Upgrade  formtastic

**Which issue(s) this PR fixes** 
https://issues.redhat.com/browse/THREESCALE-5661

**Special notes for your reviewer:**

Changelogs:
https://github.com/formtastic/formtastic/blob/3.0-stable/CHANGELOG
https://github.com/formtastic/formtastic/blob/4.0.0/CHANGELOG.md

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -68,7 +68,7 @@ gem 'acts_as_list', '~> 0.9.17'
 gem 'braintree', '~> 2.93'
 gem 'bugsnag', '~> 6.11'
 gem 'cancancan', '~> 2.3.0'
-gem 'formtastic', '~> 2.3.1'
+gem 'formtastic', '~> 4.0'
```
Comment (mayorova) [inline Gemfile:71]: Maybe `4.0.0` just to be on the safe side?...
The next [release](https://github.com/formtastic/formtastic#compatibility) will require Ruby 6.0, and we don't know what version will it be release as.

Comment (akostadinov) [reply Gemfile:71]: Bundle will not install gems that are incompatible with current ruby version.

File: app/lib/three_scale/semantic_form_builder.rb [hunk 2]
```diff
@@ -28,15 +28,6 @@ def text_field_with_errors( method, opts = {})
...
 
-    # Get a column object for a specified attribute method - if possible.
-    #
-    def column_for(method) #:nodoc:
-      # our old formtastic expects nil as in Rails 4, not a null object
-      # https://github.com/rails/rails/pull/15878
-      column = super
-      ActiveRecord::ConnectionAdapters::NullColumn === column ? nil : column
-    end
-
```
Comment (jlledom) [inline app/lib/three_scale/semantic_form_builder.rb:39]: Why removing this?

File: lib/templates/slim/scaffold/_form.html.slim [hunk 3]
```diff
@@ -0,0 +1,8 @@
+= semantic_form_for @<%= singular_name %> do |f|
+  = f.inputs do
+  <%- attributes.each do |attribute| -%>
+    = f.input :<%= attribute.name %>
+  <%- end -%>
+
+  = f.actions do
+    = f.action :submit, as: :input
```
Comment (jlledom) [inline lib/templates/slim/scaffold/_form.html.slim:8]: What is this for? is it needed?

**Review by jlledom — COMMENTED** [review]
I checked out the branch and everything seems to work fine in my local env. Also the test pass. But I'm surprised this is so easy, provided we jump from 2.X to 4.X skipping an entire major version. What do you think @josemigallas

File: app/lib/three_scale/semantic_form_builder.rb (hunk 2 above)
Comment (nidhi-soni1104) [reply app/lib/three_scale/semantic_form_builder.rb:39]: not required in new version

File: lib/templates/slim/scaffold/_form.html.slim (hunk 3 above)
Comment (nidhi-soni1104) [reply lib/templates/slim/scaffold/_form.html.slim:8]: Even I was not sure that if we really need this but pushed it to get your views

Comment (nidhi-soni1104) [reply lib/templates/slim/scaffold/_form.html.slim]: this was just to show as an example how form will look like and not needed so removed

**Comment by josemigallas** [comment]
> I checked out the branch and everything seems to work fine in my local env. Also the test pass. But I'm surprised this is so easy, provided we jump from 2.X to 4.X skipping an entire major version. What do you think @josemigallas?

I agree, this upgrade intrigues me... some methods have been removed in formtastic 3, such as `:input_html => { value: ... }` which we use, but it isn't broken, why? 🤔 According to formtastic's wiki, no changes should be required when upgrading to v3 but there's no mention of v4. 

In any case, please approach with caution @nidhi-soni1104 Maybe ask QE to run their test suite before merging this.

**Comment by mayorova** [comment]
> some methods have been removed in formtastic 3, such as `:input_html => { value: ... }`

Can you share the source please?

I also saw this [here](https://github.com/formtastic/formtastic/blob/3.0-stable/CHANGELOG#L14), but looking at the mentioned PR, it seems that it actually deprecates input's `:value` options, **in favor of** `:input_html => { value: ... }`.

**Comment by nidhi-soni1104** [comment]
> 

we have used 

> The method `find_raw_collection_for_column` contains `options[:find_options]` that was removed in Formtastic 3.0 (see [formtastic/formtastic#1025](https://github.com/formtastic/formtastic/pull/1025))
> 
> I searched for this method, and I think it's only used in another method `plan_selector_input` which is also dead code, from what it looks like.
> 
> I think the last usage of it was removed by #2364, so I think we can remove both methods.
> 
> Also, I noticed that `slider_input` is not used anymore, maybe we can drop it too.

Okay let me check all these

**Comment by nidhi-soni1104** [comment]
@mayorova PR is updated

File: Gemfile [hunk 4]
```diff
@@ -73,7 +73,8 @@ gem 'acts_as_list', '~> 0.9.17'
...
+gem 'gruff', '~>0.3', require: false
```
Comment (mayorova) [inline Gemfile:77]: Can you please rebase on master or merge master and make sure `gruff` is not included?

It was removed here https://github.com/3scale/porta/pull/3165/files

**Review by mayorova — CHANGES_REQUESTED** [review]
Checking this [changelog](https://github.com/formtastic/formtastic/blob/4.0.0/CHANGELOG.md#310rc2), I found that we are using `member_label`

Here is the deprecation warning:
```
DEPRECATION WARNING: The :member_label option is deprecated in favour of `passing an Array of label/value pairs like [['Justin', 2], ['Kate', 3]] into :collection directly (consider building the array in your model using Model.pluck)`. :member_label will be removed in the next version (called from each at {porta home}/app/views/provider/admin/user/notification_preferences/_form.html.slim:4)
```

Can you please fix it?

**Comment by nidhi-soni1104** [comment]
@mayorova all changes done.

File: app/views/provider/admin/user/notification_preferences/_form.html.slim [hunk 5]
```diff
@@ -6,11 +6,10 @@
...
 
       ol
-
         = f.input :enabled_notifications, as: :check_boxes_with_hints,
-                                          collection: category.notifications,
-                                          member_label: ->(value) { t(value, scope: :notification_preference_titles) },
-                                          member_hint: ->(value) { t(value, scope: :notification_preference_hints) }
+                                  collection: category.notifications.map { |value| [t(value, scope: :notification_preference_titles), value] },
+                                  member_hint: ->(value) { t(value, scope: :notification_preference_hints) }
+
```
Comment (mayorova) [inline app/views/provider/admin/user/notification_preferences/_form.html.slim:12]: Here is an extra line that can be removed.

Comment (nidhi-soni1104) [reply app/views/provider/admin/user/notification_preferences/_form.html.slim:12]: done

**Review by mayorova — APPROVED** [review]
Seems to be working OK :)

Please "Squash and merge"

**Pull request merged**